src/safe-upgrade-tool/developer.md

```diff
@@ -0,0 +1,150 @@
+---
+group: software-update-guide
+title: Developer information
+functional_areas:
+  - Upgrade
+---
+
+This topic contains information for developers who want to know more technical information about the Magento Safe Upgrade Tool (SUT). It is focused on developers who work closely with the Magento source code. You can use this knowledge to customize the SUT's components.
+
+## M-Ray integration
+
+M-Ray integration is an internal integration solution.
+
+The integration with M-Ray is done through:
+
+`Sut\Domain\MRay\MRayInterface`
+
+Its implementation is through the `config/services.yaml` file. Its value decides where the response of methods `api()` and `modules()` comes from.
+
+Edit this file to customize the response according to your installation. Just replace the value assigned to `Sut\Domain\MRay\MRayInterface`:
+
+### Example of a custom value
+
+`Sut\Domain\MRay\MRayInterface : "@sut_mray_mock"`
+
+In the previous example, the SUT uses `@sut_mray_mock` as the `MRayInterface` implementation. The responses from the `api()` and `modules()` methods come from the following files:
+
+*  `dev/mray_mock_files/api.json`
+*  `dev/mray_mock_files/modules.json`
+
+{:.bs-callout-info}
+When you make changes to the `services.yaml` file, delete the `var/cache/` folder to correctly apply them.
+
+## Unit testing
+
+To run the unit tests, execute one of the following commands:
+
+*  `vendor/bin/phpunit tests/unit`
+*  `vendor/bin/phpunit -c tests/unit/phpunit.xml.dist tests/unit`
+*  `vendor/bin/phpunit -c tests/unit/phpunit.xml.dist --testsuite=unit-tests`
+
+## Integration testing
+
+To run the integration tests, execute one of the following commands:
+
+*  `vendor/bin/phpunit -c tests/integration/phpunit.xml.dist tests/integration`
+*  `vendor/bin/phpunit -c tests/integration/phpunit.xml.dist --testsuite=integration-tests`
+
+## Acceptance testing
+
+1. Before executing acceptance tests, you must set the Magento URL in the `phpunit` configuration file.
+1. Copy the default `tests/acceptance/phpunit.xml` file (without the .dist suffix).
+1. Change the `TESTS_BASE_URL` value to point to the Magento URL that you want to test.
+1. To run the acceptance tests, execute one of the following commands:
+
+*  `vendor/bin/phpunit -c tests/acceptance/phpunit.xml tests/acceptance`
```

Comment:
This unordered list needs to be indented so that it aligns under the preceding ordered list.